File: src/main/scala/com/ft/notificationsmonitor/PushConnector.scala
package com.ft.notificationsmonitor

import akka.actor.{Actor, ActorRef, Props}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.headers.{Authorization, BasicHttpCredentials}
import akka.http.scaladsl.model.{HttpRequest, StatusCodes}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Sink, Source}
import com.ft.notificationsmonitor.PushConnector.{Connect, StreamEnded}
import com.ft.notificationsmonitor.PushReader.{CancelStreams, Read}
import com.ft.notificationsmonitor.model.HttpConfig
import org.slf4j.LoggerFactory

import scala.concurrent.duration._
import scala.util.{Failure, Success}

class PushConnector(private val httpConfig: HttpConfig,
                    private val pairMatcher: ActorRef) extends Actor {

  implicit private val sys = context.system
  implicit private val ec = context.dispatcher
  implicit private val mat = ActorMaterializer()

  private val logger = LoggerFactory.getLogger(getClass)
  private val connectionFlow = Http().outgoingConnectionHttps(httpConfig.hostname, httpConfig.port)
  private var reader = context.actorOf(PushReader.props(pairMatcher))
  private var cancelStreams = false

  override def receive: Receive = {
    case Connect =>
      val request = HttpRequest(uri = httpConfig.uri)
        .addHeader(Authorization(BasicHttpCredentials(httpConfig.credentials._1, httpConfig.credentials._2)))
      val responseF = Source.single(request)
        .via(connectionFlow)
        .runWith(Sink.head)
      responseF onComplete {
        case Failure(exception) =>
          logger.warn("Failed request. Retrying in a few moments... host={} uri={}", Array(httpConfig.hostname, httpConfig.uri, exception):_*)
          context.system.scheduler.scheduleOnce(5 seconds, self, Connect)
        case Success(response) =>
          if (!response.status.equals(StatusCodes.OK)) {
            logger.warn("Response status not ok. Retrying in a few moments... host={} uri={} status={}", Array(httpConfig.hostname, httpConfig.uri, response.status.intValue().toString):_*)
            context.system.scheduler.scheduleOnce(5 seconds, self, Connect)
          } else {
            logger.info("Connected to push feed. host={} uri={} status={}", Array(httpConfig.hostname, httpConfig.uri, response.status.intValue().toString):_*)
            reader = context.actorOf(PushReader.props(pairMatcher))
            reader ! Read(response.entity.dataBytes)
          }
      }

    case CancelStreams =>
      cancelStreams = true
      reader ! CancelStreams

    case StreamEnded =>
      if (!cancelStreams) {
        self ! Connect
      }
  }
}

object PushConnector {

  def props(httpConfig: HttpConfig, pairMatcher: ActorRef) = Props(new PushConnector(httpConfig, pairMatcher))

  case object Connect

  case object StreamEnded
}

File: src/main/java/com/ft/notificationsmonitor/PushConnector.java
package com.ft.notificationsmonitor;

import akka.NotUsed;
import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Creator;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.KillSwitches;
import akka.stream.Materializer;
import akka.stream.UniqueKillSwitch;
import akka.stream.javadsl.*;
import akka.util.ByteString;
import com.ft.notificationsmonitor.http.PushHttp;
import scala.concurrent.duration.Duration;

import java.util.concurrent.TimeUnit;

import static com.ft.notificationsmonitor.PushReader.*;

public class PushConnector extends UntypedActor {

    static final String CONNECT = "Connect";
    static final String RECONNECT = "Reconnect";
    static final String SHUTDOWN = "Shutdown";
    static final String READER_FAILED = "ReaderFailed";
    private static final ByteString DELIMITER = ByteString.fromString("\n\n");

    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);
    private Materializer mat = ActorMaterializer.create(getContext());

    private PushHttp pushHttp;
    private ActorRef pairMatcher;
    private ActorRef reader;
    private Cancellable heartbeatMonitor;
    private UniqueKillSwitch killSwitch;
    private boolean isShuttingDown = false;

    public PushConnector(PushHttp pushHttp, ActorRef pairMatcher) {
        this.pushHttp = pushHttp;
        this.pairMatcher = pairMatcher;
    }

    @Override
    public void onReceive(Object message) throws Throwable {
        if (message.equals(CONNECT)) {
            pushHttp.makeRequest()
                    .whenComplete((response, failure) -> {
                if (failure != null) {
                    log.error("Failed request. Retrying in a few moments...", failure);
                    getContext().system().scheduler().scheduleOnce(Duration.apply(15, TimeUnit.SECONDS), self(), CONNECT, getContext().dispatcher(), self());
                } else {
                    log.info("Connected to push feed.");
                    reader = getContext().actorOf(PushReader.props(pairMatcher), "pushReader");
                    getContext().watch(reader);
                    final Pair<UniqueKillSwitch, NotUsed> killSwitchAndDone = consumeBodyStreamInReader(response);
                    killSwitch = killSwitchAndDone.first();
                    heartbeatMonitor = getContext().system().scheduler().schedule(Duration.apply(1, TimeUnit.MINUTES),
                            Duration.apply(1, TimeUnit.MINUTES), reader, CHECK_HEARTBEAT, getContext().dispatcher(), getSelf());
                }
            });


        } else if (message.equals(RECONNECT)) {
            killSwitch.shutdown();
            heartbeatMonitor.cancel();
            reader.tell(PoisonPill$.MODULE$, self());
            getSelf().tell(CONNECT, getSelf());

        } else if (message.equals(SHUTDOWN)) {
            isShuttingDown = true;
            killSwitch.shutdown();
            heartbeatMonitor.cancel();
            reader.tell(PoisonPill$.MODULE$, self());

        } else if (message.equals(READER_FAILED)) {
            if (!isShuttingDown) {
                self().tell(RECONNECT, self());
            }

        } else if (message instanceof Terminated) {
            log.warning("{} dead.", ((Terminated) message).actor().path());
        }
    }

    private Pair<UniqueKillSwitch, NotUsed> consumeBodyStreamInReader(Source<ByteString, Object> body) {
        return body.viaMat(KillSwitches.single(), Keep.right())
                .via(Framing.delimiter(DELIMITER, 1024, FramingTruncation.ALLOW))
                .toMat(
                        Sink.actorRefWithAck(reader, INIT, ACK, COMPLETE, t -> {
                            log.error(t, "Error following push stream.");
                            return null;
                        }),
                        Keep.both()
                )
                .run(mat);
    }

    public static Props props(final PushHttp pushHttp, final ActorRef pairMatcher) {
        return Props.create(new Creator<PushConnector>() {
            private static final long serialVersionUID = 1L;

            @Override
            public PushConnector create() throws Exception {
                return new PushConnector(pushHttp, pairMatcher);
            }
        });
    }
}
